src/vaadin-number-field.html

```diff
@@ -165,22 +165,65 @@
         }
 
         _decreaseValue() {
-          this._allowed(-1) && this.__add(-1);
+          const incrementSign = this._getAllowedIncrementSign(-1);
+
+          if (incrementSign) {
+            this.__add(incrementSign);
+          }
         }
 
         _increaseValue() {
-          this._allowed(1) && this.__add(1);
+          const incrementSign = this._getAllowedIncrementSign(1);
+
+          if (incrementSign) {
+            this.__add(incrementSign);
+          }
         }
 
         __add(sign) {
           const incr = sign * (this.step || 1);
           // Behave like native number input adjusting to the next exact multiple of step.
-          this.value = this.focusElement.value =
-            (incr + (incr * Math.floor((parseFloat(this.value || 0) / incr).toFixed(1)))).toFixed(this.__decimals);
+          this.value = this.focusElement.value = this._getValue(incr);
           this.dispatchEvent(new CustomEvent('change', {bubbles: true}));
         }
 
-        _allowed(sign, value, min, max) {
+        _getValue(incr) {
+          return (incr + (incr * Math.floor((parseFloat(this.value || 0) / incr).toFixed(1)))).toFixed(this.__decimals);
+        }
+
+        _getAllowedIncrementSign(sign) {
+          if (this.disabled) {
+            return false;
+          }
+
+          const initialIncr = sign * (this.step || 1);
+
+          if (initialIncr > 0) {
+            const validMin = this.min == null || this._getValue(initialIncr) > this.min;
+            const validMax = this.max == null || this.value < this.max;
+
+            if (validMin && validMax) {
+              return sign;
+            } else if (!validMax) {
+              return false;
+            } else if (!validMin) {
+              return this.min / initialIncr;
+            }
+          } else {
+            const validMin = this.min == null || this.value > this.min;
+            const validMax = this.max == null || this._getValue(initialIncr) < this.max;
+
+            if (validMin && validMax) {
+              return sign;
+            } else if (!validMin) {
+              return false;
+            } else if (!validMax) {
+              return this.max / (this.step || 1);
+            }
+          }
+        }
+
+        _allowed(sign) {
           const incr = sign * (this.step || 1);
           return !this.disabled && (incr < 0
             ? this.min == null || this.value > this.min
```

Comment:
The check `this.value  > this.min` needs to be fixed. Currently the following scenario is not covered:

1. set value to `null`
2. set `min` to negative number, e.g. -10
3. `null > -10` here returns true
4. the "minus" control is not disabled

Reply:
I don't understand why that is expected behaviour.

1. I set value to `null` and to me it feels like `null` should behave like the value is `0` which it does.
2. If `min` is `-10` then from empty value I should be able to use the decrease button to go from empty (0) value to -10.
3. And then I should be able to click decrease button until the value is -10 at which point the decrease button is disabled

<img width="636" alt="screenshot 2019-01-23 at 16 51 54" src="https://user-images.githubusercontent.com/8615573/51614868-35ea9200-1f2f-11e9-926f-a4ade5b2176f.png">

Reply:
Discussed in person with Tomi and I agree with your point.
Let's consider `null` as 0 and implicit default value as also 0, as it is now.